File: leetcode/354-russian-doll-envelopes.py
class Solution(object):
    def maxEnvelopes(self, envelopes):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        def lower_bound(a, k, comp=operator.lt):
            l, cnt = 0, len(a)
            while cnt > 0:
                step = cnt / 2
                m = l + step
                if comp(a[m], k):
                    l = m + 1
                    cnt -= step + 1
                else:
                    cnt = step
            return l

        envelopes.sort(
            lambda a, b: b[1] - a[1] if a[0] == b[0] else a[0] - b[0])

        lis  = []
        for e in envelopes:
            p = lower_bound(lis, e[1])
            if p == len(lis):
                lis.append(e[1])
            else:
                lis[p] = e[1]
        return len(lis)

File: leetcode/354-russian-doll-envelopes.js
/**
 * @param {number[][]} envelopes
 * @return {number}
 */
var maxEnvelopes = function(envelopes) {
  function lower_bound(a, k, comp=function(a, b) { return a < b; }) {
    var l = 0, cnt = a.length;
    while (cnt > 0) {
      var step = cnt >> 1;
      var m = l + step;
      if (comp(a[m], k)) {
        l = m + 1;
        cnt -= step + 1;
      } else cnt = step;
    }
    return l;
  }

  var sorted = envelopes.sort(function(a, b) {
    return a[0] === b[0] ? b[1] - a[1] : a[0] - b[0];
  });

  var lis = [];
  for (var i = 0; i < sorted.length; ++i) {
    var p = lower_bound(lis, sorted[i][1]);
    if (p == lis.length) lis.push(sorted[i][1]);
    else lis[p] = sorted[i][1];
  }
  return lis.length;
};
